server/base/src/main/java/org/apache/accumulo/server/security/handler/KerberosAuthenticator.java

```diff
@@ -71,35 +68,30 @@ public boolean validSecurityHandlers() {
   }
 
   private void createUserNodeInZk(String principal) throws KeeperException, InterruptedException {
-    synchronized (zooCache) {
-      zooCache.clear();
-      ZooReaderWriter zoo = context.getZooSession().asReaderWriter();
-      zoo.putPrivatePersistentData(zkUserPath + "/" + principal, new byte[0],
-          NodeExistsPolicy.FAIL);
-    }
+    context.getZooCache().clear(zkUserPath + "/" + principal);
+    ZooReaderWriter zoo = context.getZooSession().asReaderWriter();
+    zoo.putPrivatePersistentData(zkUserPath + "/" + principal, new byte[0], NodeExistsPolicy.FAIL);
   }
 
   @Override
   public void initializeSecurity(String principal, byte[] token) {
     try {
       // remove old settings from zookeeper first, if any
       ZooReaderWriter zoo = context.getZooSession().asReaderWriter();
-      synchronized (zooCache) {
-        zooCache.clear();
-        if (zoo.exists(zkUserPath)) {
-          zoo.recursiveDelete(zkUserPath, NodeMissingPolicy.SKIP);
-          log.info("Removed {}/ from zookeeper", zkUserPath);
-        }
-
-        // prep parent node of users with root username
-        // ACCUMULO-4140 The root user needs to be stored un-base64 encoded in the znode's value
-        byte[] principalData = principal.getBytes(UTF_8);
-        zoo.putPersistentData(zkUserPath, principalData, NodeExistsPolicy.FAIL);
-
-        // Create the root user in ZK using base64 encoded name (since the name is included in the
-        // znode)
-        createUserNodeInZk(Base64.getEncoder().encodeToString(principalData));
+      context.getZooCache().clear((path) -> path.startsWith(zkUserPath));
```

Comment:
It looks like these synchronized blocks were dependent on previous ZooCache implementation that synchronized the reads from ZooCache. That is no longer the case, so these should be revisited as a follow-on. This code is old and I'm not sure what it was trying to do clearing the ZooCache. Presumably it was so any permissions/auth changes were immediately consistent within the same process. However, if that's the case, then clear needs to take place after the writes to ZooKeeper, not before, since ZooCache read methods are no longer synchronized.

This is probably a bug in all branches, and should be fixed in 2.1 and later.